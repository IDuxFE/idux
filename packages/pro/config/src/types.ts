/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ProLocale } from '@idux/pro/locales'
import type { ProTableColumnIndexable } from '@idux/pro/table'

export interface ProGlobalConfig {
  common: ProCommonConfig
  locale: ProLocale

  table: ProTableConfig
  tree: ProTreeConfig
}

export type ProGlobalConfigKey = keyof ProGlobalConfig

export interface ProCommonConfig {
  prefixCls: string
}

export interface ProTableConfig {
  columnIndexable: Omit<ProTableColumnIndexable, 'type'>
}

export interface ProTreeConfig {
  clearIcon: string
  collapseIcon: [string, string]
}
