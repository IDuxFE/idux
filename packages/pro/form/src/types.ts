/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import type { AbstractControl, AsyncValidatorFn, ValidatorFn } from '@idux/cdk/forms'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray } from '@idux/cdk/utils'
import type { FormItemProps, FormProps } from '@idux/components/form'
import type { RowProps } from '@idux/components/grid'
import type { Options as AjvOptions } from 'ajv'
import type { Component, DefineComponent, HTMLAttributes, PropType, VNodeChild } from 'vue'

export const proFormProps = {
  ajvOptions: Object as PropType<AjvOptions>,
  autoId: { type: [Boolean, String] as PropType<boolean | string>, default: undefined },
  autoLabelFor: { type: Boolean, default: undefined },
  fields: Object as PropType<ProFormFieldsSchema>,
  ignoreKeywords: Array as PropType<string[]>,
  schema: Object as PropType<ProFormJsonSchema>,
  schemaFormatter: Function as PropType<ProFormSchemaFormatter>,

  //events
  onSubmit: [Function, Array] as PropType<MaybeArray<(formData: any) => void>>,
} as const

export type ProFormProps = ExtractInnerPropTypes<typeof proFormProps>
export type ProFormPublicProps = ExtractPublicPropTypes<typeof proFormProps>
export type ProFormComponent = DefineComponent<
  Omit<HTMLAttributes, keyof ProFormPublicProps> & ProFormPublicProps & FormProps
>
export type ProFormInstance = InstanceType<DefineComponent<ProFormProps>>

type ArrayElement<A> = A extends (infer T)[] ? T : never

export type ProFormSchemaFormatter = (
  fields: ProFormFieldsSchema | undefined,
  schema: ProFormJsonSchema | undefined,
) => { fields: ProFormFieldsSchema; schema: ProFormJsonSchema }

// custom render(slots)
export interface ProFormItemCustoms {
  customControl?: string | ((props: { control: AbstractControl; [key: string]: any }) => VNodeChild)
  customControlTooltip?: string | (() => VNodeChild)
  customDescription?: string | (() => VNodeChild)
  customLabel?: string | (() => VNodeChild)
  customLabelTooltip?: string | (() => VNodeChild)
  customMessage?: string | (() => VNodeChild)
}

export interface ProFormFieldsSchema<T = any> extends ProFormItemCustoms {
  formProps?: FormProps
  formItemProps?: FormItemProps & RowProps
  component?: string | Component
  componentProps?: any

  default?: T

  disabled?: boolean
  name?: string
  example?: string
  trigger?: 'change' | 'blur' | 'submit'
  validators?: ValidatorFn | ValidatorFn[]
  asyncValidators?: AsyncValidatorFn | AsyncValidatorFn[]

  order?: string[]
  type?: 'object' | 'array'

  // show?: boolean | ((control: AbstractControl) => boolean)
  // if?: boolean | ((control: AbstractControl) => boolean)

  properties?: { readonly [K in keyof T]: ProFormFieldsSchema<T[K]> }
  items?: ProFormFieldsSchema<ArrayElement<T>>
}

export type ProFormJsonSchema<T = any> =
  | ObjectKeywords<T>
  | ArrayKeywords<T>
  | StringKeywords<T>
  | NumberKeywords<T>
  | BooleanKeywords<T>

export interface CommonKeywords<T> {
  title?: string
  description?: string
  const?: any
  enum?: Readonly<(T | null)[]>
  default?: T | null
}

export interface ObjectKeywords<T> extends CommonKeywords<T> {
  type?: 'object'
  properties: { readonly [K in keyof T]: ProFormJsonSchema<T[K]> }
  required?: (keyof T)[]
}

export interface ArrayKeywords<T> extends CommonKeywords<T> {
  type: 'array'
  items?: ProFormJsonSchema<ArrayElement<T>>
  minItems?: number
  maxItems?: number
}

export interface StringKeywords<T> extends CommonKeywords<T> {
  type: 'string'
  minLength?: number
  maxLength?: number
  pattern?: string
  format?: string
}

export interface NumberKeywords<T> extends CommonKeywords<T> {
  type: 'number' | 'integer'
  minimum?: number
  maximum?: number
  format?: string
}

export interface BooleanKeywords<T> extends CommonKeywords<T> {
  type: 'boolean'
}
