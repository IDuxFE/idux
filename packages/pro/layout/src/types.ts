/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { IxInnerPropTypes, IxPublicPropTypes, VKey } from '@idux/cdk/utils'
import type { LayoutSiderProps } from '@idux/components/layout'
import type { MenuClickOptions, MenuData, MenuProps, MenuTheme } from '@idux/components/menu'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export interface SiderHover {
  delay: number
}

export interface SiderHoverCtrl extends SiderHover {
  enable: boolean
}

export const proLayoutProps = {
  activeKey: IxPropTypes.oneOfType<VKey>([String, Number, Symbol]),
  collapsed: IxPropTypes.bool,
  siderHover: IxPropTypes.oneOfType([IxPropTypes.bool, IxPropTypes.object<SiderHover>()]),
  compress: IxPropTypes.bool.def(true),
  fixed: IxPropTypes.oneOfType([Boolean, IxPropTypes.object<{ header: boolean; sider: boolean }>()]).def(false),
  menus: IxPropTypes.array<MenuData>().def(() => []),
  sider: IxPropTypes.object<LayoutSiderProps>(),
  siderMenu: IxPropTypes.object<MenuProps>(),
  theme: IxPropTypes.oneOfType([
    IxPropTypes.oneOf(['light', 'dark']),
    IxPropTypes.object<{ header: MenuTheme; sider: MenuTheme }>(),
  ]).def('light'),
  type: IxPropTypes.oneOf<ProLayoutType>(['header', 'sider', 'both', 'mixin']).def('mixin'),

  // event
  'onUpdate:activeKey': IxPropTypes.emit<(activeKey: VKey | null) => void>(),
  'onUpdate:collapsed': IxPropTypes.emit<(collapsed: boolean) => void>(),
  onMenuClick: IxPropTypes.emit<(options: MenuClickOptions) => void>(),
}

export type ProLayoutProps = IxInnerPropTypes<typeof proLayoutProps>
export type ProLayoutPublicProps = IxPublicPropTypes<typeof proLayoutProps>
export type ProLayoutComponent = DefineComponent<
  Omit<HTMLAttributes, keyof ProLayoutPublicProps> & ProLayoutPublicProps
>
export type ProLayoutInstance = InstanceType<DefineComponent<ProLayoutProps>>

export type ProLayoutFixed = boolean | { header: boolean; sider: boolean }
export type ProLayoutType = 'header' | 'sider' | 'mixin' | 'both'

export const proLayoutSiderTriggerProps = {
  icon: IxPropTypes.arrayOf(IxPropTypes.oneOfType([String, IxPropTypes.vNode])),
}

export type ProLayoutSiderTriggerProps = IxInnerPropTypes<typeof proLayoutSiderTriggerProps>
export type ProLayoutSiderTriggerPublicProps = IxPublicPropTypes<typeof proLayoutSiderTriggerProps>
export type ProLayoutSiderTriggerComponent = DefineComponent<
  Omit<HTMLAttributes, keyof ProLayoutSiderTriggerPublicProps> & ProLayoutSiderTriggerPublicProps
>
export type ProLayoutSiderTriggerInstance = InstanceType<DefineComponent<ProLayoutSiderTriggerProps>>
