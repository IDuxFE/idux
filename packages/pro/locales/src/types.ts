/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

export interface ProTableLocale {
  layout: {
    title: string
    sm: string
    md: string
    lg: string
    placeholder: string
    all: string
    reset: string
    indexable: string
    expandable: string
    selectable: string
    startPin: string
    endPin: string
    noPin: string
    startPinTitle: string
    endPinTitle: string
    noPinTitle: string
  }
}

export interface ProTreeLocale {
  expandAll: string
  collapseAll: string
}

export interface ProSearchLocale {
  keyword: string
  ok: string
  cancel: string
  selectAll: string
  placeholder: string
  switchToTimePanel: string
  switchToDatePanel: string
}

export interface ProLocale {
  type: ProLocaleType

  table: ProTableLocale
  tree: ProTreeLocale
  search: ProSearchLocale
}

export type ProLocaleType = 'zh-CN' | 'en-US'
