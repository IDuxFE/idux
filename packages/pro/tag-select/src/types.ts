/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { AbstractControl, ValidateStatus } from '@idux/cdk/forms'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray, VKey } from '@idux/cdk/utils'
import type { FormSize } from '@idux/components/form'
import type { HeaderProps } from '@idux/components/header'
import type { OverlayContainerType } from '@idux/components/utils'
import type { DefineComponent, HTMLAttributes, PropType, Slot, VNode, VNodeChild } from 'vue'

export interface TagSelectColor {
  key: VKey
  name: string
  labelColor: string
  backgroundColor: string
  borderColor?: string
}

export interface TagSelectData {
  key: VKey
  label: string
  color: VKey | TagSelectColor
  disabled?: boolean

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: any
}

export const proTagSelectProps = {
  control: {
    type: [String, Number, Object, Array] as PropType<string | number | (string | number)[] | AbstractControl>,
    default: undefined,
  },
  value: { type: Array as PropType<VKey[]>, default: undefined },
  open: { type: Boolean, default: undefined },
  colors: Array as PropType<TagSelectColor[]>,
  clearable: { type: Boolean, default: false },
  clearIcon: { type: String, default: undefined },
  dataSource: Array as PropType<TagSelectData[]>,
  dataEditable: { type: Boolean, default: true },
  disabled: { type: Boolean, default: false },
  confirmBeforeSelect: { type: [Boolean, String] as PropType<boolean | 'force'>, default: false },
  maxTags: { type: [Number, String] as PropType<number | 'responsive'>, default: Number.MAX_SAFE_INTEGER },
  tagsLimit: { type: Number, default: Number.MAX_SAFE_INTEGER },
  tagDataLimit: { type: Number, default: Number.MAX_SAFE_INTEGER },
  overlayClassName: { type: String, default: undefined },
  overlayContainer: {
    type: [String, HTMLElement, Function] as PropType<OverlayContainerType>,
    default: undefined,
  },
  overlayMatchWidth: { type: [Boolean, String] as PropType<boolean | 'minWidth'>, default: 'minWidth' },
  placeholder: { type: String, default: undefined },
  readonly: { type: Boolean, default: false },
  createdTagDataModifier: Function as PropType<(data: TagSelectData) => TagSelectData>,
  removeConfirmHeader: [String, Object] as PropType<string | HeaderProps>,
  removeConfirmTitle: [String, Object, Function] as PropType<string | VNode | (() => VNodeChild)>,
  selectConfirmHeader: [String, Object] as PropType<string | HeaderProps>,
  beforeSelectConfirm: Function as PropType<(data: TagSelectData) => boolean | Promise<boolean>>,
  beforeRemoveConfirm: Function as PropType<(data: TagSelectData) => boolean | Promise<boolean>>,
  size: { type: String as PropType<FormSize>, default: undefined },
  status: String as PropType<ValidateStatus>,
  suffix: { type: String, default: undefined },

  'onUpdate:value': [Function, Array] as PropType<MaybeArray<(value: VKey[] | undefined) => void>>,
  'onUpdate:open': [Function, Array] as PropType<MaybeArray<(open: boolean) => void>>,
  onClear: [Function, Array] as PropType<MaybeArray<(evt: Event) => void>>,
  onChange: [Function, Array] as PropType<
    MaybeArray<(value: VKey[] | undefined, oldValue: VKey[] | undefined) => void>
  >,
  onFocus: [Function, Array] as PropType<MaybeArray<(evt: FocusEvent) => void>>,
  onBlur: [Function, Array] as PropType<MaybeArray<(evt: FocusEvent) => void>>,
  onTagDataChange: [Function, Array] as PropType<MaybeArray<(data: TagSelectData) => void>>,
  onTagDataRemove: [Function, Array] as PropType<MaybeArray<(data: TagSelectData) => void>>,
  onTagDataAdd: [Function, Array] as PropType<MaybeArray<(data: TagSelectData) => void>>,
  onTagSelect: [Function, Array] as PropType<MaybeArray<(data: TagSelectData) => void>>,
  onTagSelectConfirm: [Function, Array] as PropType<MaybeArray<(data: TagSelectData) => void>>,
  onTagSelectCancel: [Function, Array] as PropType<MaybeArray<(data: TagSelectData) => void>>,
  onTagRemove: [Function, Array] as PropType<MaybeArray<(data: TagSelectData) => void>>,
} as const

export interface ProTagSelectSlots {
  clearIcon: Slot
  suffix: Slot
  selectedLabel: Slot<TagSelectData>
  overflowedLabel: Slot<TagSelectData[]>
  tagLabel: Slot<TagSelectData>
  optionLabel: Slot<TagSelectData>
  maxExceededAlert: Slot
  placeholder: Slot

  selectConfirmContent: Slot<TagSelectData>
  removeConfirmTitle: Slot<TagSelectData>
  removeConfirmContent: Slot<TagSelectData>
}

export interface ProTagSelectBindings {
  blur: () => void
  focus: (options?: FocusOptions) => void
}

export type ProTagSelectProps = ExtractInnerPropTypes<typeof proTagSelectProps>
export type ProTagSelectPublicProps = ExtractPublicPropTypes<typeof proTagSelectProps>
export type ProTagSelectComponent = DefineComponent<
  Omit<HTMLAttributes, keyof ProTagSelectPublicProps> & ProTagSelectPublicProps
>
export type ProTagSelectInstance = InstanceType<DefineComponent<ProTagSelectProps>>
