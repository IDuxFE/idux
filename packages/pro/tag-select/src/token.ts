/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { OperationsContext } from './composables/useOperations'
import type { OverlayStateContext } from './composables/useOverlayState'
import type { PanelActiveStateContext } from './composables/usePanelActiveState'
import type { MergedTagData, TagDataContext } from './composables/useTagData'
import type { TagEditContext } from './composables/useTagEdit'
import type { ProTagSelectProps, TagSelectColor } from './types'
import type { VKey } from '@idux/cdk/utils'
import type { SelectorInstance } from '@idux/components/selector'
import type { ProTagSelectLocale } from '@idux/pro/locales'
import type { ComputedRef, InjectionKey, Ref } from 'vue'

export interface ProTagSelectContext
  extends TagDataContext,
    TagEditContext,
    OperationsContext,
    OverlayStateContext,
    PanelActiveStateContext {
  triggerRef: Ref<SelectorInstance | undefined>
  props: ProTagSelectProps
  focused: ComputedRef<boolean>
  mergedTagSelectColors: ComputedRef<TagSelectColor[]>
  selectedValue: ComputedRef<VKey[] | undefined>
  mergedData: ComputedRef<MergedTagData[]>
  maxExceeded: ComputedRef<boolean>
  inputValue: ComputedRef<string | undefined>
  inputFullyMatched: ComputedRef<boolean>
  mergedPrefixCls: ComputedRef<string>
  locale: ProTagSelectLocale
}

export const proTagSelectContext: InjectionKey<ProTagSelectContext> = Symbol('proTagSelectContext')
