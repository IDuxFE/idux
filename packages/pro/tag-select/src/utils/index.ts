/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export interface Deferred<V = any> {
  wait: () => Promise<V>
  resolve: (value: V) => void
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function createDeferred<V = any>(): Deferred<V> {
  let resolve: (value: V) => void
  const promise = new Promise<V>(_resolve => {
    resolve = _resolve
  })

  return {
    wait: () => promise,
    resolve: value => resolve(value),
  }
}
