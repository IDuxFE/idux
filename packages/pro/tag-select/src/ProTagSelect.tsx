/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, normalizeClass, provide, ref, watch } from 'vue'

import { useAccessorAndControl } from '@idux/cdk/forms'
import { VKey, callEmit, useState } from '@idux/cdk/utils'
import { type ControlTriggerSlots, IxControlTrigger } from '@idux/components/control-trigger'
import { useFormItemRegister, useFormStatus } from '@idux/components/form'
import { IxSelector, type SelectorInstance } from '@idux/components/selector'
import { useMergedCommonControlProps } from '@idux/components/utils'
import { useGlobalConfig } from '@idux/pro/config'
import { useThemeToken } from '@idux/pro/theme'

import { useKeyboardEvents } from './composables/useKeyboardEvents'
import { useOperations } from './composables/useOperations'
import { useOverlayState } from './composables/useOverlayState'
import { usePanelActiveState } from './composables/usePanelActiveState'
import { useRemoveConfirm } from './composables/useRemoveConfirm'
import { useSelectConfirm } from './composables/useSelectConfirm'
import { useSelectedState } from './composables/useSelectedState'
import { useTagColors } from './composables/useTagColors'
import { useTagData } from './composables/useTagData'
import { useTagEdit } from './composables/useTagEdit'
import SelectedTag from './content/SelectedTag'
import TagDataRemoveConfirmModal from './content/TagDataRemoveConfirmModal'
import TagSelectPanel from './content/TagSelectPanel'
import { proTagSelectContext } from './token'
import { proTagSelectProps } from './types'
import { getThemeTokens } from '../theme'

export default defineComponent({
  name: 'IxProTagSelect',
  inheritAttrs: false,
  props: proTagSelectProps,
  setup(props, { expose, slots, attrs }) {
    const commonConfig = useGlobalConfig('common')
    const { globalHashId, hashId, registerToken } = useThemeToken('proTagSelect')
    registerToken(getThemeTokens)

    const locales = useGlobalConfig('locale')
    const config = useGlobalConfig('tagSelect')
    const { accessor, control } = useAccessorAndControl()
    const commonControlProps = useMergedCommonControlProps(props, config)
    useFormItemRegister(control)
    const mergedStatus = useFormStatus(props, control)

    const [focused, setFocused] = useState(false)

    const mergedPrefixCls = computed(() => `${commonConfig.prefixCls}-tag-select`)

    const triggerRef = ref<SelectorInstance>()
    const focus = () => triggerRef.value?.focus()
    const blur = () => triggerRef.value?.blur()

    const tagColorsContext = useTagColors(props, locales.tagSelect)
    const tagDataContext = useTagData(props, tagColorsContext)
    const overlayStateContext = useOverlayState(props)
    const tagEditContext = useTagEdit(overlayStateContext)
    const selectedStateContext = useSelectedState(props, accessor, tagDataContext)
    const selectConfirmContext = useSelectConfirm(props, tagDataContext, selectedStateContext, overlayStateContext)
    const removeConfirmContext = useRemoveConfirm(tagDataContext, selectedStateContext, overlayStateContext)
    const operationContext = useOperations(
      tagDataContext,
      selectConfirmContext,
      removeConfirmContext,
      selectedStateContext,
    )
    const panelActiveStateContext = usePanelActiveState(tagDataContext, selectedStateContext)
    const handleKeydown = useKeyboardEvents(
      props,
      tagDataContext,
      selectedStateContext,
      selectConfirmContext,
      overlayStateContext,
      panelActiveStateContext,
      operationContext,
    )

    const { mergedData, inputValue, setInputValue } = tagDataContext
    const { selectedData, selectedValue } = selectedStateContext
    const { dataToSelect, handleTagSelectCancel } = selectConfirmContext
    const { dataToEdit } = tagEditContext
    const { overlayOpened, setOverlayOpened, setEditPanelOpened } = overlayStateContext
    const { handleTagClear } = operationContext

    const mergedSelectedData = computed(() => {
      const data = [...selectedData.value]

      if (dataToSelect.value) {
        data.push(dataToSelect.value)
      }

      return data
    })

    watch(mergedData, data => {
      if (
        (focused.value || overlayOpened.value) &&
        dataToEdit.value &&
        data.findIndex(item => item.key === dataToEdit.value?.key) < 0
      ) {
        setEditPanelOpened(false)
        setTimeout(() => {
          triggerRef.value?.focus()
        })
      }
    })
    provide(proTagSelectContext, {
      triggerRef,
      props,
      focused,
      mergedPrefixCls,
      locale: locales.tagSelect,
      mergedTagSelectColors: tagColorsContext.mergedTagSelectColors,
      maxExceeded: selectedStateContext.maxExceeded,
      selectedValue,
      ...tagDataContext,
      ...tagEditContext,
      ...operationContext,
      ...overlayStateContext,
      ...panelActiveStateContext,
    })

    expose({ focus, blur })

    const clearInput = () => {
      triggerRef.value?.clearInput()
    }

    const onFocus = (evt: FocusEvent) => {
      setFocused(true)
      callEmit(props.onFocus, evt)
    }
    const onBlur = (evt: FocusEvent) => {
      setFocused(false)
      clearInput()

      if (props.confirmBeforeSelect !== 'force' && dataToSelect.value) {
        handleTagSelectCancel()
      }

      accessor.markAsBlurred()
      callEmit(props.onBlur, evt)
    }

    const overlayClasses = computed(() => {
      const { overlayClassName } = props
      const prefixCls = mergedPrefixCls.value
      return normalizeClass({
        [globalHashId.value]: !!globalHashId.value,
        [hashId.value]: !!hashId.value,
        [`${prefixCls}-overlay`]: true,
        [overlayClassName || '']: !!overlayClassName,
      })
    })

    const renderSelectorItem = (options: {
      disabled?: boolean
      prefixCls: string
      removable?: boolean
      value?: VKey
    }) => <SelectedTag {...options} v-slots={slots} />

    const renderTrigger: ControlTriggerSlots['trigger'] = ({
      borderless,
      status,
      clearable,
      clearIcon,
      readonly,
      disabled,
      size,
      suffix,
      suffixRotate,
      focused,
      opened,
    }) => [
      <IxSelector
        ref={triggerRef}
        v-slots={{
          selectedItem: renderSelectorItem,
          placeholder: slots.placeholder,
          suffix: slots.suffix,
          clearIcon: slots.clearIcon,
        }}
        class={mergedPrefixCls.value}
        input={inputValue.value}
        allowInput={true}
        borderless={borderless}
        clearable={clearable}
        clearIcon={clearIcon}
        dataSource={mergedSelectedData.value}
        disabled={disabled}
        focused={focused}
        maxLabel={props.maxTags}
        multiple={true}
        opened={opened}
        placeholder={props.placeholder}
        readonly={readonly}
        size={size}
        status={status}
        suffix={suffix}
        suffixRotate={suffixRotate}
        onClear={handleTagClear}
        onInputValueChange={setInputValue}
      />,
    ]

    const renderContent: ControlTriggerSlots['overlay'] = () => {
      return [<TagSelectPanel v-slots={slots} />, <TagDataRemoveConfirmModal v-slots={slots} />]
    }

    return () => {
      const { suffix, borderless, clearIcon, size } = commonControlProps.value
      const controlTriggerProps = {
        autofocus: false,
        overlayClassName: overlayClasses.value,
        overlayContainer: props.overlayContainer,
        overlayContainerFallback: `${mergedPrefixCls.value}-overlay-container`,
        overlayMatchWidth: props.overlayMatchWidth,
        class: [mergedPrefixCls.value, globalHashId.value, hashId.value],
        borderless,
        value: selectedValue.value,
        open: overlayOpened.value,
        readonly: props.readonly,
        size,
        status: mergedStatus.value,
        suffix,
        clearable: props.clearable,
        clearIcon,
        disabled: accessor.disabled,
        onKeydown: handleKeydown,
        onFocus,
        onBlur,
        'onUpdate:open': setOverlayOpened,
      }

      return (
        <IxControlTrigger
          v-slots={{ trigger: renderTrigger, overlay: renderContent }}
          {...controlTriggerProps}
          {...attrs}
        />
      )
    }
  },
})
