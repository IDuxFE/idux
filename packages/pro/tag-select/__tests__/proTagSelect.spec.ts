import { MountingOptions, mount } from '@vue/test-utils'

import { renderWork } from '@tests'

import ProTagSelect from '../src/ProTagSelect'
import { ProTagSelectProps } from '../src/types'

describe.skip('ProTagSelect', () => {
  const ProTagSelectMount = (options?: MountingOptions<Partial<ProTagSelectProps>>) =>
    mount(ProTagSelect, { ...(options as MountingOptions<ProTagSelectProps>) })

  renderWork<ProTagSelectProps>(ProTagSelect, {
    props: {},
  })

  test('xxx work', async () => {
    const wrapper = ProTagSelectMount({ props: { xxx: 'Xxx' } })

    expect(wrapper.classes()).toContain('ix-Xxx')

    await wrapper.setProps({ xxx: 'Yyy' })

    expect(wrapper.classes()).toContain('ix-Yyy')
  })
})
