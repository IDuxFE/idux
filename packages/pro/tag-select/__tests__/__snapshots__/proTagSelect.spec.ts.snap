// Vitest Snapshot v1

exports[`ProTagSelect > render work 1`] = `
"<div class=\\"ix-trigger ix-trigger-md ix-trigger-paddingless ix-selector ix-selector-md ix-selector-allow-input ix-selector-multiple ix-pro-tag-select ix-control-trigger ix-pro-tag-select\\" tabindex=\\"-1\\">
  <!---->
  <div class=\\"ix-selector-content\\">
    <div class=\\"ix-selector-placeholder\\">
      <!---->
    </div>
    <div class=\\"ix-overflow ix-selector-overflow \\">
      <!---->
      <div class=\\"ix-overflow-item ix-overflow-suffix\\">
        <div class=\\"ix-selector-input\\"><input class=\\"ix-selector-input-inner\\" style=\\"opacity: 0; width: 0px;\\" readonly=\\"\\"><span class=\\"ix-selector-input-mirror\\" aria-hidden=\\"true\\"></span></div>
      </div>
    </div>
  </div>
  <div class=\\"ix-trigger-suffix\\"><i class=\\"ix-icon ix-icon-down\\" style=\\"transform: rotate(0deg);\\" role=\\"img\\" aria-label=\\"down\\"></i></div>
  <!---->
</div>"
`;
